rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1089 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+
+### 1.1. GRU
+
+GRU is defined by the following formula:
+
+![GRU-LBR](formula-gru-lbr.png)
+
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+## 2. API Changes
+
+Attention memory descriptor can be built from existing `rnn_desc_t` fields (memory
+descriptors, cell kind, direction, propagation kind, etc.). Taking into account
+this fact and to avoid breaking changes an attention memory descriptor will not
+be part of `dnnl_rnn_desc_t`. Instead a user should use query API to get
+attention shapes.
+
+ * Pros:
+   - RNN primitive descriptor can request the best layout for an attention memory
+   - RNN primitive descriptor can request the data type compatible with RNN cell
+ * Cons:
+   - User has to reorder/quantize their attention data to align with the one
+   requested by RNN
+ * Alternative:
+   - Introduce `rnn_desc_v2_t` with a memory descriptor for an attention.
+
+
+### 2.1. New Types
+
+```c
+/// Source argument #3.
+#define DNNL_ARG_SRC_3 4
+/// A special mnemonic for RNN input recurrent cell attention vector. An alias for
+/// #DNNL_ARG_SRC_3.
+#define DNNL_ARG_SRC_LAYER_ATTENTION DNNL_ARG_SRC_3
```

Comment:
Why SRC_LAYER in the name? Why not only `DNNL_ARG_AUGRU_ATTENTION`?

Reply:
This is a good question. The reason is that I left DNNL_ARG_AUGRU_ATTENTION for classic attention used in seq2seq models in case they are combined (imagine seq2seq with augru cell where 2 kinds of attention are in place). So this SRC_LAYER is to specify that this attention is a part of cell and is passed from the bottom and not from left/right side of a network.

Reply:
Not sure I follow, why would we use AUGRU in the name for classic sea2seq attention vectors? If we want to support attention vector for regular seq2seq models it makes to total sense to specify SRC_LAYER or SRC_ITER in the name since the goal would be to avoid a concat layer.

But in the case of AUGRU, there is no notion of concatenation. So using `AUGRU` in the name should makes it non ambiguous already.

Reply:
My concern was that there are 2 types of attention:
- The one which is used in AUGRU and which is applied to the intermediate result. This attention has size [T, N, 1], so it is already computed for each timestamp.
- An attention used in seq2seq models. When input for each new timestamp of RNN decoder depends on an attention mechanism.

- An idea was to split those 2 in case we will face an RNN which uses both of them.
- Another point was to keep it aligned with SRC arguments.

It is a minor detail and I am fine to split it from src proceed with `DNNL_ARG_AUGRU_ATTENTION`